src/treelearner/cuda/cuda_histogram_constructor.cu

```diff
@@ -0,0 +1,597 @@
+            cuda_row_data_->cuda_partition_ptr_uint16(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
+        } else if (cuda_row_data_->row_ptr_bit_type() == 32) {
+          CUDAConstructHistogramSparseKernel_GlobalMemory<uint16_t, uint32_t><<<grid_dim, block_dim, 0, cuda_stream_>>>(
+            cuda_smaller_leaf_splits,
+            cuda_gradients_, cuda_hessians_,
+            cuda_row_data_->cuda_data_uint16(),
+            cuda_row_data_->cuda_row_ptr_uint32(),
+            cuda_row_data_->cuda_partition_ptr_uint32(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
+        } else if (cuda_row_data_->row_ptr_bit_type() == 64) {
+          CUDAConstructHistogramSparseKernel_GlobalMemory<uint16_t, uint64_t><<<grid_dim, block_dim, 0, cuda_stream_>>>(
+            cuda_smaller_leaf_splits,
+            cuda_gradients_, cuda_hessians_,
+            cuda_row_data_->cuda_data_uint16(),
+            cuda_row_data_->cuda_row_ptr_uint64(),
+            cuda_row_data_->cuda_partition_ptr_uint64(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
+        }
+      } else if (cuda_row_data_->bit_type() == 32) {
+        if (cuda_row_data_->row_ptr_bit_type() == 16) {
+          CUDAConstructHistogramSparseKernel_GlobalMemory<uint32_t, uint16_t><<<grid_dim, block_dim, 0, cuda_stream_>>>(
+            cuda_smaller_leaf_splits,
+            cuda_gradients_, cuda_hessians_,
+            cuda_row_data_->cuda_data_uint32(),
+            cuda_row_data_->cuda_row_ptr_uint16(),
+            cuda_row_data_->cuda_partition_ptr_uint16(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
+        } else if (cuda_row_data_->row_ptr_bit_type() == 32) {
+          CUDAConstructHistogramSparseKernel_GlobalMemory<uint32_t, uint32_t><<<grid_dim, block_dim, 0, cuda_stream_>>>(
+            cuda_smaller_leaf_splits,
+            cuda_gradients_, cuda_hessians_,
+            cuda_row_data_->cuda_data_uint32(),
+            cuda_row_data_->cuda_row_ptr_uint32(),
+            cuda_row_data_->cuda_partition_ptr_uint32(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
+        } else if (cuda_row_data_->row_ptr_bit_type() == 64) {
+          CUDAConstructHistogramSparseKernel_GlobalMemory<uint32_t, uint64_t><<<grid_dim, block_dim, 0, cuda_stream_>>>(
+            cuda_smaller_leaf_splits,
+            cuda_gradients_, cuda_hessians_,
+            cuda_row_data_->cuda_data_uint32(),
+            cuda_row_data_->cuda_row_ptr_uint64(),
+            cuda_row_data_->cuda_partition_ptr_uint64(),
+            cuda_row_data_->cuda_partition_hist_offsets(),
+            num_data_,
+            cuda_hist_buffer_);
```

Comment:
maybe use macros to reduce the duplicated codes above, here and everywhere.

Reply:
Resovled via 6a9d530